tfx/orchestration/experimental/optimizations/beam_fusion.py

```diff
@@ -0,0 +1,197 @@
+          if parent in current_subgraph:
+            is_fuseable = False
+            break
+          next_layer.append(parent)
+
+      current_layer = next_layer
+
+    return is_fuseable
+
+  def get_fuseable_subgraphs(self):
+    """Returns a list of fuseable Beam component subgraphs in topological order.
+
+    Conducts multiple BFS searches to build out fuseable Apache Beam component
+    subgraphs. Each subgraph S must meet the following correctness constraints:
+    1. Each subgraph must be a directed acyclic graph
+    2. All the components in the subgraph must be Apache Beam Components
+    3. Consider a component in a subgraph being explored. None of the
+       component's ancestors (other than its' immediate parents and those
+       parents' ancestors) should be in the current subgraph being explored.
+
+    Additionally, each subrgaph must meet the following optimality constraint:
+    1. There can be no subgraphs such that their union satisfies 1, 2, and 3
+       as described above.
+    """
+
+    # Finds subgraphs of fuseable beam components through a BFS of each source.
+    candidate_sources = self.get_subgraph_sources(self.pipeline.components)
+    fuseable_subgraphs = []
+    visited = set()
+
+
+    for source in sorted(candidate_sources, key=lambda c: c.id): # pylint: disable=too-many-nested-blocks
+      if source in visited:
+        continue
+
+      subgraph = []
+      current_layer = [source]
+
+      # Conduct a BFS on the candidate source.
+      while current_layer:
+        next_layer = []
+        for component in current_layer:
+          if component in visited:
+            continue
+
+          visited.add(component)
+          subgraph.append(component)
+
+          # Iterate through the children to find Beam components
+          for child in component.downstream_nodes:
+            if not self._is_beam_component(child):
+              continue
+
+            # Checks if the child is in an explored subgraph that needs to be
+            # fused into the current subrgaph
+            if child in visited:
+              intersecting_subgraph = self._get_intersecting_subgraph(
+                  child, fuseable_subgraphs)
+              if intersecting_subgraph:
+                subgraph.extend(intersecting_subgraph)
+                fuseable_subgraphs.remove(intersecting_subgraph)
+
+            elif self._is_fuseable(child, subgraph):
+              next_layer.append(child)
+
+          current_layer = next_layer
+
+      if len(subgraph) > 1:
+        fuseable_subgraphs.append(subgraph)
+
+    # Topologically sort all the subgraphs
+    for i in range(len(fuseable_subgraphs)): # pylint: disable=consider-using-enumerate
+      fuseable_subgraphs[i] = self._topologically_sort(fuseable_subgraphs[i],
+                                                       candidate_sources)
+    return fuseable_subgraphs
+
+  def get_subgraph_sources(self, subgraph: List[base_node.BaseNode]):
```

Comment:
Although `get_subgraph_sources()` and `get_subgraph_sinks()` are not required to get `get_fuseable_subgraphs()` to work, I added them because we will need their functionality when we replace the subgraph with a `FusedComponent`. Also, `get_subgraph_sources()`  is called once in `get_fuseable_subgraphs()`  to set `candidate_sources`. Both methods are also tested in the unit tests.